battlepass/src/lib.rs

```diff
@@ -0,0 +1,257 @@
+pub type Resource<T> = BoundedVec<
+	ResourceInfoMin<
+		BoundedVec<u8, <T as Config>::StringLimit>,
+		BoundedVec<PartId, <T as Config>::PartsLimit>,
+	>,
+	<T as Config>::MaxResourcesOnMint,
+>;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+  
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_rmrk_core::Config {
+		type Event: From<Event<Self>>
+			+ IsType<<Self as frame_system::Config>::Event>
+			+ Into<<Self as frame_system::Config>::Event>;
+
+		type Control: ControlTrait<Self::AccountId, Self::Hash>;
+
+		type Rmrk: Collection<String<Self>, Symbol<Self>, Self::AccountId>
+			+ Nft<Self::AccountId, String<Self>, Resource<Self>>;
+
+		/// The maximum length of a name, cid or metadata strings stored on-chain.
+		#[pallet::constant]
+		type StringLimit: Get<u32>;
+
+		/// The maximum length of a Collection symbol.
+		#[pallet::constant]
+		type SymbolLimit: Get<u32>;
+
+		/// The maximum number of parts each resource may have
+		#[pallet::constant]
+		type PartsLimit: Get<u32>;
+
+		type MaxResourcesOnMint: Get<u32>;
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// New BattlePass created
+		BattlepassCreated {
+			org_id: T::Hash,
+			battlepass_id: T::Hash,
+			season: u32
+		},
+		
+		/// BattlePass claimed
+		BattlepassClaimed {
+			claimer: T::AccountId,
+			org_id: T::Hash,
+			battlepass_id: T::Hash
+		},
+
+		/// BattlePass closed
+		BattlepassClosed {
+			closed_by: T::AccountId,
+			org_id: T::Hash,
+			battlepass_id: T::Hash
+		},
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		AuthorizationError,
+		OrgPrimeUnknown,
+		OrgUnknown,
+		BattlepassExists,
+		BattlepassClaimed,
+		BattlepassUnknown,
+		NotMember,
+		CollectionUnknown,
+	  }
+
+	/// Battlepass by its id.
+	///
+	/// Battlepasses: map Hash => Battlepass
+	#[pallet::storage]
+	#[pallet::getter(fn get_battlepass)]
+	pub(super) type Battlepasses<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, Battlepass<T::Hash, T::AccountId, T::BlockNumber, String<T>>, OptionQuery>;
+
+	/// Battlepass state.
+	///
+	/// BattlepassStates: map Hash => BattlepassState
+	#[pallet::storage]
+	#[pallet::getter(fn get_battlepass_state)]
+	pub type BattlepassStates<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, BattlepassState, OptionQuery>;
+
+	/// Total number of battlepasses per organization.
+	///
+	/// BattlepassCount: map Hash => u32
+	#[pallet::storage]
+	#[pallet::getter(fn get_battlepass_count)]
+	pub type BattlepassCount<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, u32, ValueQuery>;
+
+	/// Current active battlepass in organization.
+	///
+	/// ActiveBattlepassByOrg: map Hash => Hash
+	#[pallet::storage]
+	#[pallet::getter(fn get_active_battlepass)]
+	pub type ActiveBattlepassByOrg<T: Config> = StorageMap<_, Blake2_128Concat, T::Hash, T::Hash, OptionQuery>;
```

Comment:
This storage has no usage.
Also, if the battlepass has 1-1 relation with the org, can we use the org's hash for the battlepass?

Reply:
Good idea. But in this case, we need to keep in mind to recreate all Battlepasses if something changed in Battlepass struct.